lib/pubsub_client.rb

```diff
@@ -1,26 +1,48 @@
 require 'pubsub_client/version'
 require 'pubsub_client/null_publisher_factory'
+require 'pubsub_client/null_subscriber_factory'
 require 'pubsub_client/publisher_factory'
+require 'pubsub_client/subscriber_factory'
 
 module PubsubClient
   Error = Class.new(StandardError)
   ConfigurationError = Class.new(Error)
   CredentialsError = Class.new(Error)
   InvalidTopicError = Class.new(Error)
+  InvalidSubscriptionError = Class.new(Error)
 
   class << self
     def stub!
-      raise ConfigurationError, 'PubsubClient is already configured' if @publisher_factory
+      raise ConfigurationError, 'PubsubClient is already configured' if @publisher_factory || @subscriber_factory
       @publisher_factory = NullPublisherFactory.new
+      @subscriber_factory = NullSubscriberFactory.new
     end
 
+    # @param message [String]
+    # @param topic [String]
     def publish(message, topic, &block)
-      unless ENV['GOOGLE_APPLICATION_CREDENTIALS']
-        raise CredentialsError, 'GOOGLE_APPLICATION_CREDENTIALS must be set'
-      end
+      ensure_credentials!
 
       @publisher_factory ||= PublisherFactory.new
       @publisher_factory.build(topic).publish(message, &block)
     end
+
+    # @param subscription [String]
+    # @param concurency [Integer]
+    # @param auto_ack [Boolean]
```

Comment:
You should expand these doc comments with descriptions of each. It may seem obvious to you today, but it won't be to others or to yourself necessarily if you have to revisit this code next year.